src/compiler/checker.ts

```diff
@@ -28040,26 +28040,31 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         }
 
         function narrowTypeBySwitchOnTrue(type: Type, switchStatement: SwitchStatement, clauseStart: number, clauseEnd: number): Type {
-            const clauses = switchStatement.caseBlock.clauses.slice(clauseStart, clauseEnd);
-            const clausesType = narrowTypeForTrueClauses(type, clauses);
             const defaultIndex = findIndex(switchStatement.caseBlock.clauses, clause => clause.kind === SyntaxKind.DefaultClause);
             const hasDefaultClause = clauseStart === clauseEnd || (defaultIndex >= clauseStart && defaultIndex < clauseEnd);
-            if (hasDefaultClause) {
-                // If we have a default in this set of clauses, then the type could also be any of the types
-                // that aren't covered by the other cases.
-
-                const clausesBefore = switchStatement.caseBlock.clauses.slice(0, clauseStart);
-                const clausesAfter = switchStatement.caseBlock.clauses.slice(clauseEnd);
 
-                const before = narrowTypeForTrueClauses(type, clausesBefore);
-                const after = narrowTypeForTrueClauses(type, clausesAfter);
-                const other = getUnionType([before, after]);
-
-                const typeNotOther = filterType(type, t => !isTypeSubsetOf(t, other));
+            // First, narrow away all of the cases that preceded this set of cases.
+            const clausesBefore = switchStatement.caseBlock.clauses.slice(0, clauseStart);
+            for (const clause of clausesBefore) {

```

Comment:
Might as well turn this into a traditional loop

Reply:
Yay, time for some off by one bugs

Reply:
I mean you don't have to, but it'd be nice to avoid the slices.

Reply:
I figured someone ask for this in review, so I already had it stashed. It really irritates me that engines can't optimize away the copy when the result is never written to...